[JSC] Store Wasm::Instance* in |codeBlock| slot

Constellation opened the pull request:
#### c9c2f3401b35fdb0f2ab6d548fbc77bc69c8e590
<pre>
[JSC] Store Wasm::Instance* in |codeBlock| slot
<a href="https://bugs.webkit.org/show_bug.cgi?id=250822">https://bugs.webkit.org/show_bug.cgi?id=250822</a>
rdar://104410328

Reviewed by Mark Lam.

This patch fixes wasm calling convention problems. The new one becomes much simpler and robust (and fixing existing bugs).

1. We remove vm.wasmContext.instance field. Previously, it was randomly configured and a lot code is missing the update of this thing (for example, wasm context switching
   inside wasm -&gt; wasm call missed the update). We should not have this kind of &quot;global&quot; variable and instead we should query to CallFrame about Wasm::Instance* of the current CallFrame*.
2. Each wasm function stores Wasm::Instance* to the CallFrame. We use |codeBlock| slot for that purpose. Since it is next to |callee| slot, we can use storePairPtr in ARM64, no code size increase.
   Plus, because we are already writing |callee| slot, it does not add new performance problem. This change makes CallFrame::lexicalGlobalObjectFromWasmCallee super simple since we can just get
   this slot from CallFrame instead of getting it from vm.wasmContext.instance. And it also makes Interpreter::unwind much simpler since we no longer need to book-keep this variable in unwinding case.
   And this also contributes to code size reduction in Wasm IC since we no longer need to have store and load code for vm.wasmContext.instance.
3. We use |codeBlock| slot for Wasm::Instance*. However, Wasm LLInt already used it for Wasm::Callee* (wasm&apos;s codeblock). This patch allocates 2 internal slots for Wasm::LLInt code and use
   this slot for Wasm::Callee* instead. This slot is &quot;WasmCodeBlock&quot;.
4. We revise the use of |this| slot for wasm functions. Previously, it was randomly used. But we use this slot to anchor JSWebAssemblyInstance from conservative GC root.
   While we are keeping Wasm::Instance* in the stack, it is not GC-managed cell. To keep Wasm functions alive while running, we need to anchor JSWebAssemblyInstance* from conservative GC root.
   We use this slot in three cases. (1) Calling wasm function from JS world / C++ world so that we need to keep wasm function alive. (2) We wasm-tail-call to a new function. Since tail-call can wipe the previous
   frame, if it is the entrance frame created by (1), we miss the anchor. Conservatively, we always store this cell in tail-call case. (3) And we do this when calling wasm function indirectly.
   This could switch to a new wasm instance, so we should keep a new instance anchored from the stack.

This change wipes a lot of weird things in wasm and makes calling convention much simpler.

* Source/JavaScriptCore/interpreter/Interpreter.cpp:
(JSC::UnwindFunctor::operator() const):
* Source/JavaScriptCore/interpreter/ProtoCallFrame.h:
* Source/JavaScriptCore/llint/WebAssembly.asm:
* Source/JavaScriptCore/wasm/WasmAirIRGeneratorBase.h:
(JSC::Wasm::ExpressionType&gt;::AirIRGeneratorBase):
* Source/JavaScriptCore/wasm/WasmB3IRGenerator.cpp:
(JSC::Wasm::B3IRGenerator::insertConstants):
* Source/JavaScriptCore/wasm/WasmCallingConvention.h:
* Source/JavaScriptCore/wasm/WasmIRGeneratorHelpers.h:
(JSC::Wasm::emitCatchPrologueShared):
* Source/JavaScriptCore/wasm/js/JSToWasm.cpp:
(JSC::Wasm::createJSToWasmWrapper):
* Source/JavaScriptCore/wasm/js/WebAssemblyFunction.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):

Canonical link: <a href="https://commits.webkit.org/259139@main">https://commits.webkit.org/259139@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e93a5c3fd69eda95cfbeed48636f44eb883842c9

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/104062 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/13157 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/36991 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/113274 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/173581 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/14217 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/4069 "Built successfully") | [❌ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/96298 "Failed to checkout and rebase branch from PR 8815") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/112354 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/109832 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/10959 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/94020 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/36/builds/96298 "Failed to checkout and rebase branch from PR 8815") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/92807 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/25632 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/36/builds/96298 "Failed to checkout and rebase branch from PR 8815") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/94138 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/6526 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/27008 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/90646 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/83/builds/4305 "Built successfully and passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/6671 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/3548 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/29901 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/12668 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/62/builds/46543 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/99257 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/8448 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/24970 "Passed tests") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/3336 "Built successfully and passed tests") | | | | 
<!--EWS-Status-Bubble-End-->

MenloDorian commented on Source/JavaScriptCore/interpreter/CallFrame.cpp:
```diff
@@ -351,9 +351,10 @@ void CallFrame::convertToStackOverflowFrame(VM& vm, CodeBlock* codeBlockToKeepAl
+JSGlobalObject* CallFrame::lexicalGlobalObjectFromWasmCallee(VM&) const
 {
-    return vm.wasmContext.load()->owner<JSWebAssemblyInstance>()->globalObject();
+    Wasm::Instance* instance = bitwise_cast<Wasm::Instance*>(const_cast<CallFrame*>(this)->uncheckedR(CallFrameSlot::codeBlock).asanUnsafePointer());
```
Instead of expressing it this way, let's add a `CallFrame::wasmInstance()` method that does:
1. ASSERT that the callFrame is a wasm callFrame.
2. Retrieves and casts the pointer in |codeBlock| as a `Wasm::Instance*`.

Constellation replied:
That sounds good! Done.

MenloDorian commented on Source/JavaScriptCore/interpreter/Interpreter.cpp:
```diff
@@ -605,9 +605,9 @@ class UnwindFunctor {
             Wasm::Callee* wasmCallee = callee.asWasmCallee();
             if (wasmCallee->hasExceptionHandlers()) {
-                JSWebAssemblyInstance* jsInstance = jsCast<JSWebAssemblyInstance*>(m_callFrame->thisValue());
+                Wasm::Instance* instance = bitwise_cast<Wasm::Instance*>(m_callFrame->uncheckedR(CallFrameSlot::codeBlock).asanUnsafePointer());
```
Ditto: use `CallFrame::wasmInstance()` here.

Constellation replied:
Done.

MenloDorian commented on Source/JavaScriptCore/interpreter/ProtoCallFrame.h:
```diff
@@ -69,6 +69,13 @@ struct JS_EXPORT_PRIVATE ProtoCallFrame {
+    {
+        codeBlockValue = bitwise_cast<CallFrame*>(instance);
+    }
+#endif
```
Let's call this `setWasmInstance` instead to make it clear that it's designed to work this way, and that this is not a hack / error.

Constellation replied:
Done.

MenloDorian commented on Source/JavaScriptCore/llint/WebAssembly.asm:
```diff
@@ -22,9 +22,10 @@
+const CalleeSaveSpaceAsVirtualRegisters = constexpr Wasm::numberOfLLIntCalleeSaveRegisters + constexpr Wasm::numberOfLLIntInternalRegisters
 const CalleeSaveSpaceStackAligned = (CalleeSaveSpaceAsVirtualRegisters * SlotSize + StackAlignment - 1) & ~StackAlignmentMask
 const WasmEntryPtrTag = constexpr WasmEntryPtrTag
+const WasmCodeBlock = CallerFrame - MachineRegisterSize - MachineRegisterSize - MachineRegisterSize
```
Why not express this as `CallerFrame - 3 * MachineRegisterSize`?  Interesting.  You're saying that the WasmCodeBlock will always be stored in effectively "local variable" space in the Wasm frame.

Also, why not define in terms of the `Wasm::numberOfLLIntCalleeSaveRegisters`?  If `Wasm::numberOfLLIntCalleeSaveRegisters` changes, won't this break?

Constellation replied:
Changed it to `CallerFrame - CalleeSaveSpaceStackAligned - MachineRegisterSize`.

> Also, why not define in terms of the Wasm::numberOfLLIntCalleeSaveRegisters? If Wasm::numberOfLLIntCalleeSaveRegisters changes, won't this break?

Yes, it doesn't work. We are checking size of callee-save registers with this value. This constant needs to be precisely the same to number of callee-saves.

MenloDorian commented on Source/JavaScriptCore/llint/WebAssembly.asm:
```diff
@@ -381,20 +368,15 @@ macro wasmPrologue()
-    storei CellTag, TagOffset + ThisArgumentOffset[cfr]
-end
-
+    storep wasmInstance, CodeBlock[cfr]
```
nit: maybe introduce a WasmInstance const, and use that instead?  The WasmInstance const can be defined to be the CodeBlock const.  I don't feel strongly about this either way.

Constellation replied:
I personally prefer `CodeBlock`. Because it needs to be a sibling to `Callee`, and we are strongly relying on the CallFrame layout described in `CallFrame.h` and CallFrameSlot, and they are using `CodeBlock` and `Callee`. So, `CodeBlock` name is better for readability.

kmiller68 replied:
I kinda think we should do what I discussed above with the two enums in `CallFrame.h`. Otherwise you have to remember that in the wasm calling convention CodeBlock is really instance... which people shouldn't have to do IMO.

kmiller68 replied:
At the very least we should have a comment in the places where we use `CodeBlock` clarifying that it's referring to the JS calling convention and that we're reusing that same offset.

MenloDorian commented on Source/JavaScriptCore/llint/WebAssembly.asm:
```diff
@@ -481,20 +463,15 @@ if WEBASSEMBLY_B3JIT and not ARMv7
-    storei CellTag, TagOffset + ThisArgumentOffset[cfr]
-end
-
+    storep wasmInstance, CodeBlock[cfr]
```
Ditto: use WasmInstance constant?

Constellation replied:
I prefer CodeBlock due to readability / consistency with the rest of the code and comments.

MenloDorian commented on Source/JavaScriptCore/llint/WebAssembly.asm:
```diff
@@ -381,20 +368,15 @@ macro wasmPrologue()
     loadp [ws1], ws1
     addp ws1, ws0
-    storep ws0, CodeBlock[cfr]
+    storep ws0, WasmCodeBlock[cfr]
```
Where's the `sp` adjustment that makes room on the stack for this WasmCodeBlock?  You should adjust `sp` before storing to the stack.

Constellation replied:
`preserveCalleeSavesUsedByWasm` is adjusting it.

MenloDorian commented on Source/JavaScriptCore/llint/WebAssembly.asm:
```diff
@@ -481,20 +463,15 @@ if WEBASSEMBLY_B3JIT and not ARMv7
     loadp [ws1], ws1
     addp ws1, ws0
-    storep ws0, CodeBlock[cfr]
+    storep ws0, WasmCodeBlock[cfr]
```
Ditto: need to adjust `sp` first to make room on the stack before storing this.

Constellation replied:
Ditto. `preserveCalleeSavesUsedByWasm` adjusted it already.

MenloDorian commented on Source/JavaScriptCore/wasm/WasmAirIRGeneratorBase.h:
```diff
@@ -3197,9 +3199,13 @@ auto AirIRGeneratorBase<Derived, ExpressionType>::addCall(uint32_t functionIndex
+            if (isTailCall) {
+                // In tail-call, we always configure JSWebAssemblyInstance* in |this| to anchor it from conservative GC roots.
+                CCallHelpers::Address thisSlot(CCallHelpers::stackPointerRegister, CallFrameSlot::thisArgument * static_cast<int>(sizeof(Register)) - prologueStackPointerDelta());
+                jit.loadPtr(CCallHelpers::Address(m_wasmContextInstanceGPR, Instance::offsetOfOwner()), GPRInfo::nonPreservedNonArgumentGPR1);
```
Ditto.

MenloDorian replied:
Nevermind this.  I was thinking that `m_wasmContextInstanceGPR` hasn't been changed to the target function's Wasm::Instance, but that's not necessary for a tail call.  This comment was left over from before I understood this, and should have been deleted.

Constellation replied:
Thanks!

MenloDorian commented on Source/JavaScriptCore/wasm/WasmAirIRGeneratorBase.h:
```diff
@@ -3267,6 +3273,10 @@ auto AirIRGeneratorBase<Derived, ExpressionType>::addCall(uint32_t functionIndex
         return { };
     }
 
+    // We do not need to store |this| with JS instance since,
```
I suggest rephrasing this as "We do not need to store the JS instance in |this|"

Constellation replied:
Changed.

MenloDorian commented on Source/JavaScriptCore/wasm/WasmAirIRGeneratorBase.h:
```diff
@@ -3267,6 +3273,10 @@ auto AirIRGeneratorBase<Derived, ExpressionType>::addCall(uint32_t functionIndex
 
+    // We do not need to store |this| with JS instance since,
+    // 1. It is not tail-call. So this does not clobber the arguments of this function.
+    // 2. We are not changing instance. Thus, |this| of this function's arguments are the same and OK.
```
I suggest rephrasing this to give a bit more detail as follows: "... are the same and will continue to provide a GC root for the JS instance."

Constellation replied:
Done

MenloDorian commented on Source/JavaScriptCore/wasm/WasmB3IRGenerator.cpp:
```diff
@@ -1090,17 +1092,13 @@ void B3IRGenerator::insertConstants()
-    m_constantInsertionValues.insertValue(0, storeThis);
-    m_constantInsertionValues.insertValue(0, storeCallSiteIndex);
+    if (storeCallSiteIndex)
+        m_constantInsertionValues.insertValue(0, storeCallSiteIndex);
```
`storeCallSiteIndex` can only be non-null if `mayHaveExceptionHandlers` is true.  So, why not just do this insertion as part of the `if (mayHaveExceptionHandlers)` statement?

Also, `storeCallSiteIndex` can be a local var in that if block since it's not needed anywhere else.

Constellation replied:
Changed.

MenloDorian commented on Source/JavaScriptCore/wasm/js/WebAssemblyFunction.cpp:
Can we also rename `instance` to `jsInstance` (maybe in a follow up patch)?  It's pretty confusing to use the same `instance` name to mean different things in different places.

Constellation replied:
Done.

MenloDorian commented on Source/JavaScriptCore/wasm/js/WebAssemblyFunction.cpp:
Let's use `setWasmInstance` instead (see comment in ProtoCallFrame.h above)?

Constellation replied:
Changed.

MenloDorian reviewed (approved):
r=me

Constellation reviewed (commented):
Thanks!

kmiller68 commented on Source/JavaScriptCore/interpreter/CallFrameInlines.h:
```diff
@@ -78,6 +78,14 @@ inline JSGlobalObject* CallFrame::lexicalGlobalObject(VM& vm) const
+inline Wasm::Instance* CallFrame::wasmInstance() const
+{
+    ASSERT(callee().isWasm());
+    return bitwise_cast<Wasm::Instance*>(const_cast<CallFrame*>(this)->uncheckedR(CallFrameSlot::codeBlock).asanUnsafePointer());
```
I almost think we should have a second enum mirroring `CallFrameSlot` that has all the wasm slots. But we can do that another day. Alternatively, we could rename this to `CallFrameSlot::jsCodeBlockOrWasmInstance` but I think having two enums is clearer.

kmiller68 commented on Source/JavaScriptCore/wasm/WasmAirIRGeneratorBase.h:
```diff
@@ -1012,17 +1011,21 @@ AirIRGeneratorBase<Derived, ExpressionType>::AirIRGeneratorBase(const ModuleInfo
+                CCallHelpers::Address calleeSlot { GPRInfo::callFrameRegister, CallFrameSlot::callee * sizeof(Register) };
+                jit.storePtr(scratchGPR, calleeSlot.withOffset(PayloadOffset));
                 jit.store32(CCallHelpers::TrustedImm32(JSValue::WasmTag), calleeSlot.withOffset(TagOffset));
+                jit.storePtr(m_wasmContextInstanceGPR, CCallHelpers::addressFor(CallFrameSlot::codeBlock));
```
Ditto, on the comment here.

kmiller68 commented on Source/JavaScriptCore/wasm/WasmIRGeneratorHelpers.h:
```diff
@@ -196,8 +196,7 @@ static inline void emitCatchPrologueShared(B3::Air::Code& code, CCallHelpers& ji
 
-    jit.loadPtr(CCallHelpers::Address(GPRInfo::callFrameRegister, CallFrameSlot::thisArgument * sizeof(Register)), GPRInfo::regT3);
-    jit.loadPtr(CCallHelpers::Address(GPRInfo::regT3, JSWebAssemblyInstance::offsetOfInstance()), GPRInfo::regT3);
+    jit.loadPtr(CCallHelpers::Address(GPRInfo::callFrameRegister, CallFrameSlot::codeBlock * sizeof(Register)), GPRInfo::regT3);
```
Ditto on comment here.

kmiller68 commented on Source/JavaScriptCore/wasm/js/WebAssemblyFunction.cpp:
```diff
@@ -386,9 +377,16 @@ CodePtr<JSEntryPtrTag> WebAssemblyFunction::jsCallEntrypointSlow()
 #if USE(JSVALUE32_64)
+    jit.storePtr(scratchJSR.payloadGPR(), CCallHelpers::addressFor(CallFrameSlot::callee));
     jit.store32(CCallHelpers::TrustedImm32(JSValue::WasmTag), CCallHelpers::addressFor(CallFrameSlot::callee).withOffset(TagOffset));
+    jit.storePtr(pinnedRegs.wasmContextInstancePointer, CCallHelpers::addressFor(CallFrameSlot::codeBlock));
```
Ditto on comment.

kmiller68 commented on Source/JavaScriptCore/wasm/js/JSToWasm.cpp:
```diff
@@ -299,8 +295,7 @@ std::unique_ptr<InternalFunction> createJSToWasmWrapper(CCallHelpers& jit, Calle
 
-        jit.loadPtr(CCallHelpers::Address(GPRInfo::callFrameRegister, JSCallingConvention::instanceStackOffset), wasmContextInstanceGPR);
-        jit.loadPtr(CCallHelpers::Address(wasmContextInstanceGPR, JSWebAssemblyInstance::offsetOfInstance()), wasmContextInstanceGPR);
+        jit.loadPtr(CCallHelpers::addressFor(CallFrameSlot::codeBlock), wasmContextInstanceGPR);
```
Ditto on comment.

kmiller68 commented on Source/JavaScriptCore/wasm/js/JSToWasm.cpp:
```diff
@@ -276,14 +277,9 @@ std::unique_ptr<InternalFunction> createJSToWasmWrapper(CCallHelpers& jit, Calle
-        jit.loadPtr(CCallHelpers::Address(wasmContextInstanceGPR, JSWebAssemblyInstance::offsetOfInstance()), wasmContextInstanceGPR);
-        ASSERT(currentInstanceGPR != InvalidGPRReg);
-
+        jit.loadPtr(CCallHelpers::addressFor(CallFrameSlot::codeBlock), wasmContextInstanceGPR);
```
Ditto on comment.

Constellation committed:
[JSC] Store Wasm::Instance* in |codeBlock| slot
https://bugs.webkit.org/show_bug.cgi?id=250822
rdar://104410328

Reviewed by Mark Lam.

This patch fixes wasm calling convention problems. The new one becomes much simpler and robust (and fixing existing bugs).

1. We remove vm.wasmContext.instance field. Previously, it was randomly configured and a lot code is missing the update of this thing (for example, wasm context switching
   inside wasm -> wasm call missed the update). We should not have this kind of "global" variable and instead we should query to CallFrame about Wasm::Instance* of the current CallFrame*.
2. Each wasm function stores Wasm::Instance* to the CallFrame. We use |codeBlock| slot for that purpose. Since it is next to |callee| slot, we can use storePairPtr in ARM64, no code size increase.
   Plus, because we are already writing |callee| slot, it does not add new performance problem. This change makes CallFrame::lexicalGlobalObjectFromWasmCallee super simple since we can just get
   this slot from CallFrame instead of getting it from vm.wasmContext.instance. And it also makes Interpreter::unwind much simpler since we no longer need to book-keep this variable in unwinding case.
   And this also contributes to code size reduction in Wasm IC since we no longer need to have store and load code for vm.wasmContext.instance.
3. We use |codeBlock| slot for Wasm::Instance*. However, Wasm LLInt already used it for Wasm::Callee* (wasm's codeblock). This patch allocates 2 internal slots for Wasm::LLInt code and use
   this slot for Wasm::Callee* instead. This slot is "WasmCodeBlock".
4. We revise the use of |this| slot for wasm functions. Previously, it was randomly used. But we use this slot to anchor JSWebAssemblyInstance from conservative GC root.
   While we are keeping Wasm::Instance* in the stack, it is not GC-managed cell. To keep Wasm functions alive while running, we need to anchor JSWebAssemblyInstance* from conservative GC root.
   We use this slot in three cases. (1) Calling wasm function from JS world / C++ world so that we need to keep wasm function alive. (2) We wasm-tail-call to a new function. Since tail-call can wipe the previous
   frame, if it is the entrance frame created by (1), we miss the anchor. Conservatively, we always store this cell in tail-call case. (3) And we do this when calling wasm function indirectly.
   This could switch to a new wasm instance, so we should keep a new instance anchored from the stack.

This change wipes a lot of weird things in wasm and makes calling convention much simpler.

* Source/JavaScriptCore/interpreter/Interpreter.cpp:
(JSC::UnwindFunctor::operator() const):
* Source/JavaScriptCore/interpreter/ProtoCallFrame.h:
* Source/JavaScriptCore/llint/WebAssembly.asm:
* Source/JavaScriptCore/wasm/WasmAirIRGeneratorBase.h:
(JSC::Wasm::ExpressionType>::AirIRGeneratorBase):
* Source/JavaScriptCore/wasm/WasmB3IRGenerator.cpp:
(JSC::Wasm::B3IRGenerator::insertConstants):
* Source/JavaScriptCore/wasm/WasmCallingConvention.h:
* Source/JavaScriptCore/wasm/WasmIRGeneratorHelpers.h:
(JSC::Wasm::emitCatchPrologueShared):
* Source/JavaScriptCore/wasm/js/JSToWasm.cpp:
(JSC::Wasm::createJSToWasmWrapper):
* Source/JavaScriptCore/wasm/js/WebAssemblyFunction.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):

Canonical link: https://commits.webkit.org/259139@main
